src/hierarchical.rs

```diff
@@ -0,0 +1,244 @@
+/// * **Complete-linkage**: Currently a naive implementation with O( n^3 ) time
+///   and O( n^2 ) space complexity.
+/// * **Single-linkage**: Currently a naive implementation with O( n^3 ) time
+///   and O( n^2 ) space complexity.
+///
+/// # Examples
+///
+/// ```rust
+/// use cogset::{Agglomerative, Euclid, LinkageCriterion};
+///
+/// let data = [Euclid([0.0, 0.0]),
+///             Euclid([1.0, 0.5]),
+///             Euclid([0.2, 0.2]),
+///             Euclid([0.3, 0.8]),
+///             Euclid([0.0, 1.0])];
+///
+/// let agglomerative = Agglomerative::new(&data);
+///
+/// println!("{:?}", agglomerative.clusters(LinkageCriterion::Complete, 0.4));
+/// println!("{:?}", agglomerative.clusters(LinkageCriterion::Complete, 0.6));
+///
+/// println!("{:?}", agglomerative.dendrogram(LinkageCriterion::Complete));
+/// ```
+///
+/// <script src="https://is.gd/BFouBe"></script>
+#[derive(Clone, Debug)]
+pub struct Agglomerative<'a, P: 'a + Point> {
+    points: Vec<&'a P>,
+    distances: Vec<Vec<f64>>
+}
+
+impl<'a, P: 'a + Point> Agglomerative<'a, P> {
+    /// Pre-compute the inter-point distances of `data`.
+    pub fn new<T>(data: T) -> Agglomerative<'a, P>
+        where T: IntoIterator<Item=&'a P>
+    {
+        let points:Vec<_> = data.into_iter().collect();
+        let n = points.len();
+
+        // pre-compute distances, using as few space as possible,
+        // we assume that d(i,j) = d(j,i) and d(i,i) = 0
+        let distances = (0..n).map(|i| {
+            (i+1..n).map(|j| {
+                points[i].dist(points[j])
+            }).collect()
+        }).collect();
+
+        Agglomerative {
+            points: points,
+            distances: distances
+        }
+    }
+
+    /// Perform the actual clustering.
+    ///
+    /// Starting with each point in its own cluster successively merge two clusters
+    /// ,determined by `linkage`, until either only one is left or the maximal distance
+    /// `threshold` has been exceeded.
+    pub fn clusters(&'a self, linkage: LinkageCriterion, threshold: f64) -> Vec<Vec<&'a P>> {
+        self.merge(linkage, threshold).iter().map(|cluster| {
+            cluster.into_iter().map(|i| self.points[*i]).collect()
+        }).collect()
+    }
+
+    /// Calculate the hierarchical dendrogram for the givan linkage criterion.
+    pub fn dendrogram(&'a self, linkage: LinkageCriterion) -> Box<Dendrogram<&'a P>> {
+        let mut x = self.merge(linkage, INFINITY);
+        assert!(x.len() == 1);
+        let x = x.pop().unwrap();
+
+        self.convert(&x)
+    }
+
+    fn convert(&'a self, root: &Dendrogram<usize>) -> Box<Dendrogram<&'a P>> {
+        Box::new(match root {
+            &Dendrogram::Branch(d, ref a, ref b) =>
+                Dendrogram::Branch(d, self.convert(a), self.convert(b)),
+            &Dendrogram::Leaf(i) => Dendrogram::Leaf(self.points[i])
+        })
+    }
+
+    fn merge(&'a self, linkage: LinkageCriterion, threshold: f64) -> Vec<Box<Dendrogram<usize>>> {
```

Comment:
Each `Dendrogram` represents a cluster, that means if I use an infinite threshold while merging, I get one large `Dendrogram` for the whole set of points, i.e., one large cluster. The `Dendrogram` is just a representation of the hierarchical merging process.

If I now use a fixed threshold, I might end up with more than one cluster, since the process might stop prematurely. That means there is more than one `Dendrogram` left, each describing a cluster. Thus I have to return a `Vec` of `Dendrogram`s rather than just one `Dendrogram`, which is limited to the former case only.